File: src/main/scala/19.remove-nth-node-from-end-of-list.scala
import common.scala.ListNode

/**
  * Definition for singly-linked list.
  * class ListNode(var _x: Int = 0) {
  *   var next: ListNode = null
  *   var x: Int = _x
  * }
  */
object Q19 {
  def removeNthFromEnd(head: ListNode, n: Int): ListNode = {
    @scala.annotation.tailrec
    def toMapWithIdx(
        n: ListNode,
        i: Int,
        acc: Map[Int, ListNode]
    ): Map[Int, ListNode] =
      if (n == null) acc
      else toMapWithIdx(n.next, i + 1, acc + (i -> n))

    // 用map缓存节点的index, 只遍历一次节点
    val map = toMapWithIdx(head, 0, Map.empty[Int, ListNode])
    val idx = map.size - n
    // 头节点要特殊对待, 删除头节点的方式就是返回第二个节点-_-
    if (idx == 0) head.next
    else {
      // 其他节点的删除方式就是把前面的跟后面的接上, 跳过自己
      map(idx - 1).next = map(idx).next
      head
    }
  }
}

File: src/main/java/19.remove-nth-node-from-end-of-list.java
import common.java.ListNode;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 * 思路:
 * <pre>
 * 双指针, 先拉开n个距离, 再利用这个距离求解
 * 难点在于要处理只有单个节点的链表这样的情况, 所以要额外引入一个头
 * </pre>
 */
class Q19Java {
    public ListNode removeNthFromEnd(ListNode head, int n) {
      ListNode hd = new ListNode(0);
      hd.next = head;
      ListNode fast = hd;
      // 这里的+1其实就是把hd包括进去
      for (int i = 0; i < n + 1; i++) {
        fast = fast.next;
      }
      ListNode slow = hd;
      while (fast != null) {
        fast = fast.next;
        slow = slow.next;
      }
      slow.next = slow.next.next;
      return hd.next;
    }
}
